File: 1a.c
#include<stdio.h>
#include<stdlib.h>
#include<time.h>

void generateRandomArray(int arr[],int n){
    for (int  i = 0; i < n; i++)
    {
        arr[i]=rand()%100000;
    }
}

int linearSearch(int arr[],int target,int n){
    for(int i=0;i<n;i++){
        if(arr[i]==target){
            return 1;
        }
    }
    return 0;
}

int main(){
    int n;
    printf("Enter size of array\n");
    scanf("%d",&n);
    
    int *arr = (int *)malloc(n * sizeof(int));
    if (arr == NULL) {
        printf("Memory allocation failed\n");
        return 1;
    }
    
    generateRandomArray(arr, n);
    clock_t start =clock();
    
    for(int i=0;i<100000;i++){
        linearSearch(arr,arr[n-1],n);
    }
    clock_t end =clock();

    double time_taken = (double)(end -start) /CLOCKS_PER_SEC;
    double avg_time = time_taken/100000.0;
    printf("Total time: %f s,Average per search: %f s\n", time_taken,avg_time);

    return 0;
}

File: 1a.py
import random
import time
import matplotlib.pyplot as plt

def generate_random_array(n):
    return [random.randint(0,100000) for _ in range(n)]


def linear_search(arr, target):
    for x in arr:
        if x ==target:
            return True
    return False


n_values = [100, 500, 1000, 2000, 5000, 10000, 20000] 
times = []

for n in n_values:
    arr = generate_random_array(n)
    target = arr[-1] 

    start = time.time()
    for _ in range(10000):  
        linear_search(arr, target)
    end = time.time()

    total_time = end- start
    avg_time = total_time /10000.0
    times.append(avg_time)

    print(f"n={n}, Total time={total_time:.6f}s, Avg per search={avg_time:.10f}s")


plt.plot(n_values, times, marker='o')
plt.xlabel("Number of elements (n)")
plt.ylabel("Average time per search (s)")
plt.title("Linear Search Time vs List Size")
plt.grid(True)
plt.show()
